### ISSUE TITLE
Conda does no longer add the variant to the build string.

### ISSUE THREAD

**Issue by rkaminsk** [conda/conda-build#2875]
### Actual Behavior

I can no longer build packages for different python versions.

I have the following build matrix:

```
$ cat conda_build_config.yaml 
python:
    - 2.7
    - 3.6
```

When building, conda uses `clingo-5.3.0-hf484d3e_5.tar.bz2` as archive name for both the python 2.7 and 3.6 build. Earlier versions added py27 and py36 as prefix. I also experimented with `pin_run_as_build` or setting `build.string` manually (which then gobbles the build number) - nothing seems to work.

Things are still working nicely in another environment with conda 4.5.0 and conda-build 3.8.0.

### Expected Behavior

At least the hashes should be different when building for different python versions.

### Steps to Reproduce

The build scripts are available at <https://github.com/potassco/conda/tree/master/clingo>. In case you want to reproduce, bison and re2c have to be installed by other means because the latter is not availabel in conda.

##### `Output of conda info`

```
     active environment : base
    active env location : /home/wv/opt/conda3
            shell level : 1
       user config file : /home/kaminski/.condarc
 populated config files : 
          conda version : 4.5.2
    conda-build version : 3.10.1
         python version : 3.6.5.final.0
       base environment : /home/wv/opt/conda3  (writable)
           channel URLs : https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/free/linux-64
                          https://repo.anaconda.com/pkgs/free/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
                          https://repo.anaconda.com/pkgs/pro/linux-64
                          https://repo.anaconda.com/pkgs/pro/noarch
          package cache : /home/wv/opt/conda3/pkgs
                          /home/kaminski/.conda/pkgs
       envs directories : /home/wv/opt/conda3/envs
                          /home/kaminski/.conda/envs
               platform : linux-64
             user-agent : conda/4.5.2 requests/2.18.4 CPython/3.6.5 Linux/4.9.0-6-amd64 debian/9 glibc/2.24
                UID:GID : 2087:500
             netrc file : None
           offline mode : False

```

**Comment by msarahan** [comment]
I think you need to remove your vc dep:  https://github.com/potassco/conda/blob/master/clingo/meta.yaml#L29

You get that for free with the compiler jinja anyway.

You might want

```
build:
  skip: True  # [win and vc<14]
```

instead

**Comment by rkaminsk** [comment]
Your solution to skip building for python versions using vc<14 is nicer than what I did. But it doesn't solve my actual problem. Even after removing that line, I have still the same problem.

From the build log:
```
...
BUILD START: [u'clingo-5.3.0-hf484d3e_5.tar.bz2']
...
BUILD START: [u'clingo-5.3.0-hf484d3e_5.tar.bz2']
...
```
Everything is building correctly but the build for python 2.7 gets overridden with the build for python 3.6.

**Comment by mingwandroid** [comment]
You should remove python from your `requirements/build` section (unless you are cross-compiling and need a python that runs on the build machine to do some work).

Then you should add `python` to your `requirements/host`. That should fix your issue.

**Comment by msarahan** [comment]
Yes, what @mingwandroid says is absolutely true.  The problem is that you're mixing the new compiler stuff without doing any of the move to the host section.  More info in the docs at https://conda.io/docs/user-guide/tasks/build-packages/define-metadata.html#host

**Comment by rkaminsk** [comment]
Thanks! I was spending half a day searching for a solution but didn't see this in the docs (or any of the examples). :smile: Now it is working with the newer conda versions, too! If I read the docs correctly, I do not even have to list python under `requirements/run`, right?

**Comment by sdebionne** [comment]
I think this issue should be reopened since having python in the `requirements/build` section (python is used by the build system in my use case) gives invalid variant string with the python part missing (conda-build 3.15.1).

Also the doc need to be updated, the example in the [variant documentation](https://github.com/conda/conda-build/blob/master/docs/source/variants.rst) has python in the `requirements/build` section and that does not work (yet), see:

meta.yaml contents like:

```
package:
    name: compiled-code
    version: 1.0

requirements:
    build:
        - python {{ python }}
    run:
```

**Comment by mingwandroid** [comment]
> gives invalid variant string with the python part missing (conda-build 3.15.1).

If you use python as a build tool only then it definitely should not appear in the hash.
